### ISSUE TITLE
SQLite: "too many terms in compound SELECT" when inserting lots of records at once

### ISSUE THREAD

**Issue by dos1** [laravel/framework#25262]
SQLite has a hard limit of 500 terms in a compound SELECT (see: https://stackoverflow.com/a/9527898). The limit can be configured by the user to be lower, but not greater.

Laravel doesn't take that into account, resulting in an "too many terms in compound SELECT" error when trying to insert more than 500 entries at once.

https://github.com/laravel/framework/blob/fe68aa9ee8e6356a95259cfe984a204940ed3b22/src/Illuminate/Database/Query/Grammars/SQLiteGrammar.php#L146-L187

A possible solution would be to split up big datasets into multiple queries (at most `SQLITE_MAX_COMPOUND_SELECT`) in SQLiteGrammar's compileInsert function. Another one would be to simply wrap many inserts into a single transaction regardless of the dataset size, which should perform similarly performance-wise.

The current limit on the database can be checked with `sqlite3_limit(db, SQLITE_MAX_COMPOUND_SELECT, -1)` function, as described in https://www.sqlite.org/c3ref/limit.html

**Comment by dos1** [comment]
Just a regular batch insert of ~700 entries. The exact context can be seen here: https://github.com/dos1/movim/commit/306de37a5bcc02440381128f872ccfe1dc8da81c - where `Roster::insert` comes from `Illuminate\Database\Eloquent\Model`. The linked commit successfully workarounds this issue - splitting it to `SQLITE_MAX_COMPOUND_SELECT` entries per batch would work too.

(btw. it's not my code, I know about it only as much as I debugged, which led me to filling this bug).

**Comment by staudenmeir** [comment]
We could fix it by using the standard syntax: `INSERT INTO "table" ("column") VALUES (?), (?)`
SQLite has been supporting it since [3.7.11](http://www.sqlite.org/releaselog/3_7_11.html) (6.5 years ago).

Ideally, we would detect the SQLite version and use the legacy syntax if required. But the problem is that we don't have access to the database version inside the grammar.

Replacing the old syntax completly would be a breaking change. And you can bet that some hosting services still use ancient SQLite versions...

This is another use case for version-specific grammars (https://github.com/laravel/ideas/issues/870).

**Comment by dos1** [comment]
@staudenmeir That syntax you mention is implemented in SQLite as just a syntax sugar and falls under the same limit as the old one.

**Comment by dos1** [comment]
@laurencei Any info on why has it been closed? I just looked at the SQLiteGrammar.php file on master branch and I don't see any relevant update there.

**Comment by dos1** [comment]
@laurencei Oh, no, it was a reply to the previous comment referring to the alternative syntax to the one used now by Laravel (before 3.8.8 it was just a syntax sugar, so changing it in Laravel won't fix the issue for the older ones). So, if Laravel is fine with requiring 3.8.8+, the syntax can simply be changed, otherwise it should do a version check and construct multiple INSERT queries on older ones (or just ignore that and leave this bug for the users of older SQLite versions).

To sum up, this is a Laravel bug that happens on the newest SQLite. Any specific SQLite versions were only mentioned when discussing a way to fix it.

**Comment by dos1** [comment]
(if you're fine with requiring SQLite 3.8.8 [or 3.7.11 if you don't mind this bug still happening for some people], I should even manage to make a PR with the fix :P)

**Comment by laurencei** [comment]
...it kind of feels like a real edge case, that if someone runs into it - the easiest/obvious solution is to update to SQLite 3.8.8?

Changing the SQL statement just to handle this single case might have a performance impact for other users *not* affected by it?

**Comment by dos1** [comment]
Uhm, no, please read the whole discussion again. This happens for every SQLite version right now, it's just a fix proposed in https://github.com/laravel/framework/issues/25262#issuecomment-414836191 that will require SQLite 3.8.8 to actually fix anything.

**Comment by laurencei** [comment]
Ok... now I see.

Well - seems like the best option is submit a PR for this and see what Taylor thinks.

**Comment by dos1** [comment]
Submitted :)

**Comment by staudenmeir** [comment]
Please close the issue.
